### ISSUE TITLE
Docker layer cache is not fail-safe or concurrency-safe

### ISSUE THREAD

**Issue by flx42** [apptainer/singularity#331]
There are a few shortcomings with the current Docker layer caching code. It doesn't handle failures well, for instance if I kill a `singularity import` during a layer download, the next run will believe the layer is cached:
```
$ singularity import tensorflow docker://tensorflow/tensorflow:nightly-devel-gpu
Cache folder set to /root/.singularity/docker
Downloading layer sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4
Extracting /root/.singularity/docker/sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4.tar.gz
Downloading layer sha256:247d4da3141d6674f423f02a06b43d09c39a5cb504c8295ca007d547f368b858
^CTraceback (most recent call last):
  File "/usr/local/libexec/singularity/python/cli.py", line 241, in <module>
    main()
  File "/usr/local/libexec/singularity/python/cli.py", line 206, in main
    token=token) 
  File "/usr/local/libexec/singularity/python/docker/api.py", line 237, in get_layer
    return api_get(base,headers=token,stream=download_folder)
  File "/usr/local/libexec/singularity/python/utils.py", line 163, in api_get
    chunk = response.read(chunk_size)
  File "/usr/lib/python2.7/socket.py", line 384, in read
    data = self._sock.recv(left)
  File "/usr/lib/python2.7/httplib.py", line 612, in read
    s = self.fp.read(amt)
  File "/usr/lib/python2.7/socket.py", line 384, in read
    data = self._sock.recv(left)
  File "/usr/lib/python2.7/ssl.py", line 756, in recv
    return self.read(buflen)
  File "/usr/lib/python2.7/ssl.py", line 643, in read
    v = self._sslobj.read(len)
KeyboardInterrupt
ERROR: Aborting with RETVAL=1
```
And then the next run:
```
$ singularity import tensorflow docker://tensorflow/tensorflow:nightly-devel-gpu
Cache folder set to /root/.singularity/docker
Extracting /root/.singularity/docker/sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4.tar.gz
Extracting /root/.singularity/docker/sha256:247d4da3141d6674f423f02a06b43d09c39a5cb504c8295ca007d547f368b858.tar.gz

gzip: stdin: unexpected end of file
tar: Unexpected EOF in archive
tar: Unexpected EOF in archive
tar: Error is not recoverable: exiting now
Downloading layer sha256:ddfb14e25e00d9a2ddc711c404784f1e06c524f8d6321034042bff3223efd2a3
Extracting /root/.singularity/docker/sha256:ddfb14e25e00d9a2ddc711c404784f1e06c524f8d6321034042bff3223efd2a3.tar.gz
Downloading layer sha256:2876a6fb9de1a3da108ec078309986cc0e16d5c48ebe87f53991888623e0d655
[...]
```
It also shouldn't swallow the error like this and should abort.

Using something like `try: download(layer) except: unlink(layer)` won't work, because we also have an issue when you start 2 `singularity import` in parallel:
```
$ singularity import tensorflow docker://tensorflow/tensorflow:nightly-devel-gpu
Cache folder set to /root/.singularity/docker
Extracting /root/.singularity/docker/sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4.tar.gz
Extracting /root/.singularity/docker/sha256:247d4da3141d6674f423f02a06b43d09c39a5cb504c8295ca007d547f368b858.tar.gz

gzip: stdin: unexpected end of file
tar: Unexpected EOF in archive
tar: Unexpected EOF in archive
tar: Error is not recoverable: exiting now
Downloading layer sha256:ddfb14e25e00d9a2ddc711c404784f1e06c524f8d6321034042bff3223efd2a3
Extracting /root/.singularity/docker/sha256:ddfb14e25e00d9a2ddc711c404784f1e06c524f8d6321034042bff3223efd2a3.tar.gz
Downloading layer sha256:2876a6fb9de1a3da108ec078309986cc0e16d5c48ebe87f53991888623e0d655
[...]
```
In this case the layer was being downloaded by another `singularity import`, thus this process believed it was cached. Even without killing the first `singularity import`.

I think the solution is to download the layer to a temporary location **with a temporary filename** like `$SINGULARITY_CACHEDIR/tmp/XXXXXX.tar.gz`. Then, if download is successful, do a `os.rename` to `$SINGULARITY_CACHEDIR/docker/hash.tar.gz`. We want `os.rename` to be sure the move is atomic, so we don't want `shutil.move` and we don't wan't to download to `/tmp` since you can't do a `rename` accross file systems.

Thoughts?

**Comment by vsoch** [comment]
oh goodness, you are right - I didn't even think of that. I like the idea of starting with a temporary location and then moving to the proper cache when it's done. Should we have a check to make sure the file downloaded properly?

**Comment by gmkurtzer** [comment]
I am assuming there is a checksum for the downloaded file? ;)

If so, your logic is good. Download to a temporary location (e.g. `${NAME}.part`) and then if it passes checksum test after download, move it to `${NAME}`, and continue on to next. If we encounter a `${NAME}.part` at any time, delete it, and start over.

**Comment by flx42** [comment]
I don't think we can have access to a checksum for verification. So our best option is to assume that the file is fine if the HTTP download completes without errors. At worst, we can `unlink` the file if the archive extraction fails, but it might fail for other reasons...

@gmkurtzer: the problem with your approach is that we have to think carefully of the possible race conditions when two processes want to download to `${NAME}.part` concurrently, that's why I said a randomly generated temporary filename might be simpler, but we shouldn't screw the cleanup if it fails :)

**Comment by vsoch** [comment]
Yes +1 what @flx42 mentions about the race condition - if two processes are downloading the same file, we don't want one deleting the other! 

Wait, for the checksum, isn't the digest id (the sha256 name) the checksum?

**Comment by flx42** [comment]
> Wait, for the checksum, isn't the digest id (the sha256 name) the checksum?

I think it's a bit more complicated than that, but if you find a simple way, let me know :)
https://gist.github.com/aaronlehmann/b42a2eaf633fc949f93b#id-definitions-and-calculations

**Comment by vsoch** [comment]
Lol. Whyyy Docker... whyyyy!

**Comment by vsoch** [comment]
It's ok, let's get the current two PRs finished up and merged, then we will tackle this one. I've done a basic atomically saving function for large files in python before, and likely @flx42 has something in mind. I can give a first pass at a fix for this, and then get feedback. :turkey:

**Comment by flx42** [comment]
OK for me! I will try to take a look at it too, maybe I can get something working quickly.

**Comment by gmkurtzer** [comment]
While I didn't mention it, flock() can easily be used to defeat the dreaded race condition but with that said... The proposal of a randomly generated temporary file name would also work and having two downloads of the same layer wont cause an issue when the files get moved to their final resting place.

I'm irked there is no checksum in the metadata...

**Comment by vsoch** [comment]
Here is [flock](https://docs.python.org/3/library/fcntl.html#fcntl.flock) via python.

**Comment by bbockelm** [comment]
Beware of `flock`:

- File locking is nasty on many shared filesystems.  A hung NFS locking daemon on a remote host can cause this to hang even though the NFS server itself is responsive.
- File locking can cause lack-of-progress if something goes wrong with another process.  For example, suppose that a process on a different host successfully grabs the lock, then sits and does nothing for 45 minutes.

There are some tricks you can do (try file locking for a minute or two, then proceed without the lock) to avoid two processes both doing the initial import.

Let's get the concurrency correct and we can optimize the download later.

**Comment by flx42** [comment]
@bbockelm agreed, the priority should be something like this:
- Stop silently discarding errors during download.
- Support concurrency during download, even at the cost of  having multiple concurrents downloads of the same file.
- Implement verification of the downloaded layers (checksum).
- Try to optimize download to avoid duplicated work, we could even consider downloading all layers in parallel.

**Comment by bbockelm** [comment]
Probably the verification of the downloaded images should be a separate ticket.  @vsoch - can you write it up?

I have mixed feelings about multiple parallel downloads: TCP fairshare is there for a purpose.  Anyhow, that also could be a separate ticket.

**Comment by flx42** [comment]
It **must** be a separate ticket. :)
Parallel download could be a nice to have, sometimes ISP throttling is so bad that you should even use multiple connections **for the same file**. Also, if the image is hosted on some kind of CDN, you might be hitting different servers anyway.

**Comment by vsoch** [comment]
@bbockelm definitely - I'm going to wait on #346 so I can implement the same strategy for both docker and shub cache.
